indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -1222,17 +1197,30 @@ private void taskComplete(
     runningTasks.remove(taskStatus.getId());
 
     // Update success/failure counters
-    if(taskStatus.isSuccess()){
-      zkWorker.resetCountinouslyFailedTasksCount();
-    } else if(taskStatus.isFailure()){
-      zkWorker.incrementCountinouslyFailedTasksCount();
-    }
+    if (zkWorker != null) {
+      if (taskStatus.isSuccess()) {
+        zkWorker.resetContinuouslyFailedTasksCount();
+        if (blackListedWorkers.remove(zkWorker)) {
+          zkWorker.setBlacklistedUntil(null);
+          log.info("[%s] removed from blacklist because a task finished with SUCCESS", zkWorker.getWorker());
+        }
+      } else if (taskStatus.isFailure()) {
+        zkWorker.incrementContinuouslyFailedTasksCount();
+      }
 
-    // BlackList node if there are too many failures.
-    if(zkWorker.getCountinouslyFailedTasksCount() > config.getMaxRetriesBeforeBlacklist() &&
-            blackListedWorkers.size() <=
-                    zkWorkers.size()*(config.getMaxPercentageBlacklistWorkers()/100)){
-      blackListedWorkers.add(zkWorker);
+      // Blacklist node if there are too many failures.
+      if (zkWorker.getContinuouslyFailedTasksCount() > config.getMaxRetriesBeforeBlacklist() &&
+          blackListedWorkers.size() <= zkWorkers.size() * (config.getMaxPercentageBlacklistWorkers() / 100.0) - 1) {
```

Comment:
This entire block should be synchronized on `blackListedWorkers` or else checking "size" before calling "add" is a race that could lead us to having too many blacklisted workers.